### PR TITLE
Copy on write using weakrefs

### LINKED ISSUE #10954: Views, Copies, and the SettingWithCopyWarning Issue
As `pandas` approaches its 1.0 release, I would like to raise a concern about one aspect of the `pandas` architecture that I think is a threat to its widespread adoption: how `pandas` works with copies and views when setting values (what I will refer to here as the `SettingWithCopyWarning` issue). 

The summary of my concern is the following:

```
1. SettingWithCopyWarning is a threat to data integrity
2. It is unreasonable to expect the average user to avoid a `SettingWithCopyWarning` issue, as doing
    so requires keeping track of the plethora of factors that determine what generates a copy and what
    generates a view.
    2a. Views made sense in `numpy`, but not in `pandas`
    2b. Chain-indexing is a much more subtle problem than suggested in the `pandas` docs. 
3. Given (1) and (2), data integrity in `pandas` relies on users noticing a non-exception warning in the
    flow of their output.
4. Even aside from the threat to data integrity, this behavior is unpythonic, and likely to frustrate
    alienate lots of potential users of `pandas`. 
5. I think solutions can be found that would have only limited effects on performance for the majority of  
    users
```

Taking each of these in turn:

(1) **`SettingWithCopyWarning` is a threat to data integrity**

The fact that assignment operations do different things depending on whether the target is a view or a copy has already been recognized as a threat to the predictability of `pandas`. Indeed, the reason a warning was added is because users were consistently asking why `pandas` was doing un-anticipated things when `SettingWithCopyWarning` came into play. 

(2) **It is unreasonable to expect the average user to avoid a `SettingWithCopyWarning` issue, as doing so requires keeping track of the plethora of factors that determine what generates a copy and what generates a view.**

Figuring out when a function will return a copy and when it will return a view in `pandas` is not simple. Indeed, the `pandas` documentation doesn't even try to explain when each will occur (link http://pandas-docs.github.io/pandas-docs-travis/indexing.html?highlight=views#indexing-view-versus-copy):

> The reason for having the SettingWithCopy warning is this. Sometimes when you slice an array you will simply get a view back, which means you can set it no problem. However, even a single dtyped array can generate a copy if it is sliced in a particular way. A multi-dtyped DataFrame (meaning it has say float and object data), will almost always yield a copy. Whether a view is created is dependent on the memory layout of the array."

(2a) **Views made sense in `numpy`, but not in `pandas`**
Views entered the pandas lexicon via `numpy`. But the reason they were so useful in `numpy` is that they were _predictable_ because `numpy` arrays are always single-typed. In `pandas`, no such consistent, predictable behavior exists. 

(2b) **Chain-indexing is a much more subtle problem than suggested in the `pandas` docs**
At first glance, the `pandas` docs suggest that the `SettingWithCopyWarning` is easily avoided by avoiding chain-indexing or using `.loc`. This, I fear, is misleading for two reasos. First, the canonical example of chain indexing in the docs (`dfmi['one']['second'] = value`) seems to suggest that one can avoid chain indexing by just not falling into the trap of this kind of double slicing. The problem, however, is that these slices need not appear near one another. I know I've had trouble with code of the form:

```
df2 = dfmi['one']

# Lots of intermediate code that doesn't change dfmi or df2

df2['second'] = 5
```

Moreover, using `.loc` only solves this problem if one notices the chained indexing and attempts to fix it in one place. Just consistently using `.loc[]` (for example, in both the first and second problematic slicings above) would not solve the problem. 

(3) **Given (1) and (2), data integrity in `pandas` relies on users noticing a non-exception warning in the flow of their output.**

This seems really problematic. If a users is printing values as they go along (which CS developers may not do, but interactive casual users often do to monitor the progress of their code), these warnings are easy to miss. And that seems very dangerous. 

(4) **Even aside from the threat to data integrity, this behavior is unpythonic, and likely to frustrate alienate lots of potential users of `pandas`**

I suspect I come to `pandas` from a different perspective than many developers. I am an economist and political scientist who has gotten deeper and deeper into computer science over the past several years for instrumental purposes. As a result, I think I have a pretty good sense of how applied users approach something like `pandas`, and I can just see this peculiarity of `pandas` driving this class of users batty. I've taken a year of computer science course work and am one of the most technically trained social scientists I know, and it drives me batty.

It's also unpythonic -- the behavior of basic operators (like `=`) should not depend on the type of columns in a `DataFrame`. Python 3 changed the behavior of the `/` operator because it was felt the behavior of `/` should not do depend on whether you were working with `float`s or `int`s. Since whether functions return a view or copy is in large part (but not exclusively) a function of whether a `DataFrame` is single or multi-typed (which occurs when some columns are `floats` and some are `ints`), we have the same problem -- the operation of a basic operation (`=`) depends on data types.

In other words, if one of the aims of `pandas` is to essentially surplant `R` among applied data scientists, then I think this is a major threat to achieving that goal. 

(5) **I think solutions can be found that would have only limited effects on performance for the majority of users**
`pandas` uses views because they're so damn fast, so I understand the reluctance to drop them, but I think there are ways to minimize the performance hit. Obviously more talented developers with a better understanding of the innards of `pandas` may have better suggestions, but hopefully this can get the ball rolling. 

```
* Solution 1: Move `views` to the background.
   When a user tries to look at an object and it's possible to return a view, do so. But just never let a 
   user assign values to a view -- any time an attempt is made to set on a view, convert it to a copy
   before executing the assignment. Views will still operate in the background providing high speed 
   data access in read-only environments, but users don't have to worry about what they're dealing 
   with. Users who *really* need access to views can work with `numpy` arrays. 

  (I would also note that given the unpredictability of when one will get a view or copy, it's not clear to 
  me how anyone can write code that takes advantage of the behavior of views, which makes me 
  doubt there are many people for whom this would seriously impact performance or written code, but 
  I'd be happy to hear if anyone has workarounds!)

* Solution 2: Create an indexer that always returns copies (like .take(), but for axis labels). 
   This would at least give users who want to avoid views all together a way to do so without littering
   their code with `.copy()`s. 

* Solution 3: Change the `SettingWithCopyWarning` to an exception by default. 
  This is currently a setting, but the global default is for it to be a warning. Personally, I still don't like 
  this solution since, as a result of (2) this means `pandas` will now raise exceptions unpredictably, but 
  at least data integrity will be preserved.     
```

`pandas` is a brilliant tool, and a huge improvement on everything else out there. I am eager to see it becomes the standard not only among python users, but among data analysts more broadly. Hopefully, by addressing this issue, we can help make this happen. 

With that in mind, I would like to suggest the need for two things:
1. A discussion about the desirability of the various solutions proposed above
2. Volunteers to help implement this change. Unfortunately, I don't have the programming sophistication or knowledge of `pandas` internals to take this on alone, and this is likely too big an undertaking for any one individual anyone, so a team is likely to be necessary.

### PR DESCRIPTION
**Pull request by nickeubank**
Working model of copy-on-write. Aims to close #10954, alternative to #10973, extension of #11207.
## Copy-on-Write Behavior:

**Setting on child doesn't affect parent, but still uses views when can for efficiency**

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
child = parent.loc[0:0,]
child._is_view

Out[1]: True

child.loc[0:0] = -88
child

Out[2]: 
   col1  col2
0   -88   -88

parent

Out[3]: 
   col1  col2
0     1     3
1     2     4
```

**Setting on parent doesn't affect child**

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
child = parent.loc[0:0,]
child._is_view

Out[4]: True

parent.loc[0:0, 'col1'] = -88
child

Out[5]: 
   col1  col2
0     1     3

parent

Out[6]: 
   col1  col2
0   -88     3
1     2     4
```

**One exception is dictionary-like access, where views are preserved**
_(as suggested here: https://github.com/pydata/pandas/issues/10954#issuecomment-136521398 )_

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
child = parent['col1']
child._is_view

Out[7]: True

parent.loc[0:0, 'col1'] = -88
child

Out[8]: 
0   -88
1     2
Name: col1, dtype: int64

parent

Out[9]: 
   col1  col2
0   -88     3
1     2     4
```

**Safe for views of views**

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
child = parent.loc[0:0,]
child._is_view
Out[20]: True

child_of_child = child.loc[0:0,'col1':'col1']
child_of_child._is_view

Out[21]: True

child_of_child.loc[0:0, 'col1'] = -88
child_of_child

Out[22]: 
   col1
0   -88

child

Out[23]: 
   col1  col2
0     1     3

parent

Out[24]: 
   col1  col2
0     1     3
1     2     4
```
## Chained indexing behavior now consistent

**Will always fail unless first class is a dictionary-like call for a single series (since that will always be a view)**

**Will fail if first call not dict-like**

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
parent.loc[0:0, 'col1']['col1'] = -88
parent

Out[10]: 
   col1  col2
0     1     3
1     2     4
```

**Will always succeed if first call dict-like**

```
parent = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
parent['col1'].loc[0:0] = -88
parent

Out[11]: 
   col1  col2
0   -88     3
1     2     4
```

To Do:
- Get feedback on implementation; 
- if sound in behavior and principles, remove all SettingWithCopy machinery
- Add docs

@jreback 
@TomAugspurger 
@shoyer 
@JanSchulz 
cc @ellisonbg
cc @CarstVaartjes

### CODE REVIEW

**Comment by shoyer** [comment]
First of all -- this is really great! Thank you for pushing on this.

A few specific design issues:

1. There are two obvious ways to do copy-on-write for views. We could copy the parent dataframe, or we could copy all the children (and their children, recursively). Right now, it looks like you're doing both? I think that's unnecessary. If I'm right, let's consider which option has more favorable performance.
2. Why are you using a list instead of a `WeakValueDictionary`? The later seems like it might be a better choice -- it's automatically cleaned up when children are garbage collected. With your current implementation, I think you are missing the check to verify that your weakrefs are valid.

Finally, we are definitely going to need some comprehensive performance checks. How does the ASV benchmark suite compare after this change? What types of uses used to be fast but now are slow? These will need to be carefully documented.

**Comment by jreback** [comment]
you need to have all of the original tests in here
some might work / be modified but you cannot eliminate tests

**Comment by nickeubank** [comment]
Thanks both for the feedback!

@shoyer 
1) In the simple case -- a parent who is not a view and a child who is a view -- there won't be any copying of the parent because of the if statement at 1235 of `generic.py`. But I'll think about weirder cases like views of views...

2) I thought about this, but wasn't sure how to key a dictionary. Thankfully, weakrefs are actually self-cleaning -- if they object they refer to gets cleaned, they redirect to `None`:

    test_df = pd.DataFrame()
    test_wr = weakref.ref(test_df)
    test_df = 'redirect_to_string'
    test_wr() is None
    Out[36]: True

But you are correct I need to add a  `if child() is None` call at the top of `_convert_views_to_copies`.

3) Could you please point me to some docs on this ASV? Never worked with it before. 

@jreback Would you mind expanding on your comment / clarifying a little? I don't think I've deleted any tests. Do you mean that you think this seems promising enough it's worth digging into the testing modules to see what works and what fails with these modifications and edit the failing tests? Happy to do so, just wanted to see if people felt this path was sufficiently promising to warrant that work.

**Comment by shoyer** [comment]
@nickeubank 

1) you're right that unnecessary copies aren't done if there are no views. My point is that it is unnecessary to copy both the parent AND its children -- only one is sufficient.

2) use `id(obj)` as the key. That's guaranteed to be unique as long as an object is in memory. The main reason I prefer using `WeakValueDictionary` to list is that the later can be a memory leak -- if you repeatedly index a DataFrame, the list will grow unbounded in size, even if the indexed dataframes themselves are garbage collected.

3) see here for ASV: http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-performance-test-suite

**Comment by jreback** [comment]
you have essentially turned off all of the old setting with copy tests
so these need to be adjusted to the new behavior
but all of the cases are valid

**Comment by nickeubank** [comment]
@jreback  are you talking about runtime tests in core files like `generic.py`, or the nosetests in the test files like `test_generic.py`?

@shoyer Got it. Revising now. Also realized one places this breaks, so need to add another tweak to implementation. Will ping when done.

One general question: Is there a better way to know if something is a view than `._is_view`? Seems to give false positives -- makes it safe, but leads to some extra copying. For example:

    df = pd.DataFrame({'col1':[1,2], 'col2':[3,4]})
    df._is_view
    Out[1]: False
    
    df.loc[0,'col1'] = -88
    df._is_view
    Out[2]: True

**Comment by nickeubank** [comment]
@shoyer @jreback Performance checks

@shoyer Think your suggestions all integrated. Still probably some excess copying because `_is_view` is conservative (sometimes calls things views when they aren't), [but not sure how to fix](http://stackoverflow.com/questions/33530753/numpy-reliable-non-conservative-indicator-if-numpy-array-is-view?noredirect=1#comment54843728_33530753) 


@jreback I'm afraid I'm still a little unclear on what you're asking for. I know you do a lot on these sites so brevity is your friend, but could you please be a little  more expansive / point to a specific example?


        before     after       ratio
      [eb66bccc] [cc6c9b1b]
    +  116.74ms      1.99s     17.00  frame_methods.frame_insert_500_columns_end.time_frame_insert_500_columns_end
    +  185.65ms      1.32s      7.10  plotting.plot_timeseries_period.time_plot_timeseries_period
    +   17.64ms   120.52ms      6.83  timeseries.timeseries_series_offset_fast.time_timeseries_series_offset_fast
    +   19.03ms   104.11ms      5.47  timeseries.timeseries_datetimeindex_offset_fast.time_timeseries_datetimeindex_offset_fast
    +   43.55ms   234.87ms      5.39  sparse.sparse_frame_constructor.time_sparse_frame_constructor
    +   45.87ms   109.18ms      2.38  frame_methods.frame_count_level_axis0_multi.time_frame_count_level_axis0_multi
    -    1.78ms   781.24μs      0.44  replace.replace_replacena.time_replace_replacena
    SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.

**Comment by jreback** [comment]
the perf issues need to be addressed

File: pandas/core/generic.py [hunk 1]
```diff
@@ -1074,13 +1079,19 @@ def get(self, key, default=None):
...
         try:
             return self[key]
         except (KeyError, ValueError, IndexError):
             return default
 
     def __getitem__(self, item):
-        return self._get_item_cache(item)
+        result = self._get_item_cache(item)
+
+        if isinstance(item, str):
```
Comment (kawochen) [inline pandas/core/generic.py]: Why only `str`?

**Comment by jreback** [comment]
@nickeubank you changed code but you never fixed tests. e.g. ``SetttingWithCopyError`` is no longer defined. So its impossible to see if what you did works or not.

You will need to carefully address each of the existing tests, as these define the current API. These must all pass and/or change as appropriate. This ensures that the API is preserved. You can certainly change things, but these must be well-defined and explictly communicated.

Generally having green PR's gives much more confidence. This is the point of tests, to try to map out the API as much as possible. To cover edge cases and to provide consistency among operations. This of course is not guaranteed, but helps a lot. Examining code only goes so far. Python is far to dynamic a language to just rely on code inspection for correctness.

**Comment by nickeubank** [comment]
@jreback Ah, ok -- thanks so much for the longer explanation! Will do!

**Comment by nickeubank** [comment]
t

**Comment by nickeubank** [comment]
@jreback thanks again for explanation on tests -- now see wisdom. Updated tests, and fixed a few issues that came up. 

**Open Issues**

1) Currently the only place views are preserved are column slices of data frames with simple indices. Is there an analogous case for panels where we can always offer views?

2) Are single column slices from multi-index DataFrames always views (i.e. do we have same guarantee as with single-index)?

3) The attribute `_is_column_view` I added in `generic.py` does not appear to be preserved when a `DataFrame` is pickled and re-loaded. Any suggestions on how to fix? Had to comment out one test for that...

**Other notes**

I removed the `copy` option from `merge`, since it's pretty inconsistent with the new copy-on-write paradigm.

**Performance**
Fixed main performance issue, still some work to do on dates.

        before     after       ratio
       [eb66bccc] [f4f8e562]
    +   14.66ms    98.57ms      6.73  timeseries.timeseries_datetimeindex_offset_fast.time_timeseries_datetimeindex_offset_fast
    +  187.66ms      1.18s      6.30  plotting.plot_timeseries_period.time_plot_timeseries_period
    +   16.61ms   100.92ms      6.07  timeseries.timeseries_series_offset_fast.time_timeseries_series_offset_fast
    +   24.86ms    61.02ms      2.46  frame_methods.frame_getitem_single_column2.time_frame_getitem_single_column2
    +   45.47ms   108.02ms      2.38  frame_methods.frame_count_level_axis0_multi.time_frame_count_level_axis0_multi
    +   26.26ms    60.99ms      2.32  frame_methods.frame_getitem_single_column.time_frame_getitem_single_column
    -    1.72ms   715.28μs      0.41  replace.replace_replacena.time_replace_replacena
    SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.

cc: @shoyer

**Comment by jreback** [comment]
2) this should work, nothing special about multiindexes
3) these attributes should **not** be pickled. Simply make them a class attribute defaulted to ``None``.

You are still removing LOTS of tests. Don't remove any! You need to individually go thru and fix/correct them. You need to assert that the entire existing API is unchanged (which is defined via the tests). Clearly if somethign is asserting that ``SettingWIthCopyWarning`` is raised it should be changed and tested that it is a view (or not a view as the case may be).

The entire point of the test suite is to avoid changing behavior accidently.

**Comment by nickeubank** [comment]
@jreback 

On 2: sorry, wasn't clear.  It isn't attribute value I want to pickle; it's the existence of the attribute. 

    In[17]:
    df = pd.DataFrame({"A": [1,2]})
    df._is_column_view
    Out[18]: False

    In[19]:
    df.to_pickle('test')
    df2=pd.read_pickle('test')
    df2._is_column_view
    Traceback (most recent call last):

     File "<ipython-input-22-18fc2f02c57c>", line 1, in <module>
        df2._is_column_view

     File "/Users/Nick/github/pandas/pandas/core/generic.py", line 2290, in __getattr__
       return object.__getattribute__(self, name)

    AttributeError: 'DataFrame' object has no attribute '_is_column_view'

Do I need to rebuild somewhere?

Will work on tests!

**Comment by jreback** [comment]
you need to define it as ``None`` on the class (e.g. notice the ``is_copy = None`` which you are removing), e.g.

```
class NDFrame(...):
     .....
     _is_column_view = None
```

then when you set it in an instance it is overriden only in that instance. This is a feature of python.

**Comment by nickeubank** [comment]
@jreback @shoyer Quick update:

Most things working well and I think will work eventually. Currently need to address:

* Issues with `__init__` in frame.py.
    + Need to prevent creation of views of `numpy` arrays. The copy-on-write machinery is all at the `pandas` level, so I can only support views of other `pandas` objects. 
    + Need to make sure if passed a DataFrame, creates an actual view, doesn't just share `._data` object (see #11855 ). 
    + For some reason, if I just set `copy=True` as default in `__init__`, getting a Segementation Fault in `test_frame.py`. Need to figure that out. 
* Need to prevent indexing operations from ever letting objects share `_data` objects instead of creating views (see #11855)
* Need to work through all tests.

However, at the moment I need to take a hiatus on this to work on a draft of my dissertation, so I'm going to have to set this aside for ~4 weeks. Haven't dropped -- I still really believe in the importance of this and that it will work -- but taking up more time than I have at the moment (especially given how much background research I'm having to learn about python to do all this...).

**Comment by jreback** [comment]
@nickeubank np, thanks for the effort so far!

**Comment by jreback** [comment]
xref #11970 

as this might take a bit, feel free to reopen

**Comment by jreback** [comment]
@nickeubank hmm, I can't seem to reopen this, very odd.

**Comment by kawochen** [comment]
because `HEAD` has changed since

**Comment by nickeubank** [comment]
k will reopen as new pr
